opentelemetry-sdk/src/trace/span.rs

```diff
@@ -198,43 +198,53 @@ impl opentelemetry::trace::Span for Span {
 
 impl Span {
     fn ensure_ended_and_exported(&mut self, timestamp: Option<SystemTime>) {
-        // skip if data has already been exported
-        let mut data = match self.data.take() {
-            Some(data) => data,
-            None => return,
+        match self.data {
+            Some(ref mut data) => {
+                // ensure end time is set via explicit end or implicitly on drop
+                if let Some(timestamp) = timestamp {
+                    data.end_time = timestamp;
+                } else if data.end_time == data.start_time {
+                    data.end_time = opentelemetry::time::now();
+                }
+            }
+            None => {
+                return;
+            }
         };
 
-        let provider = self.tracer.provider();
-        // skip if provider has been shut down
-        if provider.is_shutdown() {
+        if self.tracer.provider().is_shutdown() {
             return;
         }
 
-        // ensure end time is set via explicit end or implicitly on drop
-        if let Some(timestamp) = timestamp {
-            data.end_time = timestamp;
-        } else if data.end_time == data.start_time {
-            data.end_time = opentelemetry::time::now();
+        #[cfg(feature = "experimental_span_processor_on_ending")]
+        {
+            let provider = self.tracer.provider().clone();
+            for processor in provider.span_processors() {
+                processor.on_ending(self);
+            }
         }
 
-        match provider.span_processors() {
-            [] => {}
-            [processor] => {
-                processor.on_end(build_export_data(
-                    data,
-                    self.span_context.clone(),
-                    &self.tracer,
-                ));
-            }
-            processors => {
-                for processor in processors {
-                    processor.on_end(build_export_data(
-                        data.clone(),
-                        self.span_context.clone(),
-                        &self.tracer,
-                    ));
-                }
-            }
+        let Span {
+            data,
+            tracer,
+            span_context,
+            ..
+        } = self;
+
+        // skip if data has already been exported
+        let data = match data.take() {
+            Some(data) => data,
+            None => return,
+        };
+        let span_context: SpanContext =
```

Comment:
It would be helpful to comment around this bit to make it easier to re-grok -- "take ownership of the span_context leaving an empty context in its place; this saves us a clone" or so, right?